### PR TITLE
Proof of concept for datadog metric types

### PR DESCRIPTION
**Pull request by wryun**
Sadly, you have to pretend to be the datadog agent...

Again, haven't yet signed the CLA, so no tests (and don't merge). But interested to hear your thoughts on this more invasive change.

### CODE REVIEW

File: plugins/outputs/datadog/datadog.go [hunk 1]
```diff
@@ -136,17 +122,44 @@ func (d *Datadog) authenticatedUrl() string {
...
+// (also has magic for statsd field names)
+func buildMetrics(m telegraf.Metric) ([]*Metric, error) {
+	var datadogMetrics []*Metric
+	tags := m.Tags()
+	metricType := tags["metric_type"]
+	baseDatadogMetric := Metric{
+		Metric:   m.Name(),
+		Tags:     buildTags(tags),
+		Host:     tags["host"],
+		Interval: float64(m.Interval()) / float64(time.Second),
```
Comment (sparrc) [inline plugins/outputs/datadog/datadog.go]: you can change this to just `m.Interval().Seconds()`

File: plugins/outputs/datadog/datadog.go [hunk 2]
```diff
@@ -136,17 +122,44 @@ func (d *Datadog) authenticatedUrl() string {
...
-		var p Point
-		if err := p.setValue(v); err != nil {
-			return ms, fmt.Errorf("unable to extract value from Fields, %s", err.Error())
+// Convert a telegraf metric to datadog metrics;
+// we need a separate metric for each field.
+// (also has magic for statsd field names)
+func buildMetrics(m telegraf.Metric) ([]*Metric, error) {
+	var datadogMetrics []*Metric
+	tags := m.Tags()
+	metricType := tags["metric_type"]
```
Comment (sparrc) [inline plugins/outputs/datadog/datadog.go]: this will panic if not sending a statsd metric?

File: plugins/outputs/datadog/datadog.go [hunk 3]
```diff
@@ -136,17 +122,44 @@ func (d *Datadog) authenticatedUrl() string {
...
 		}
+		metric.Metric = strings.Replace(metric.Metric, "_", ".", -1)
+		p := &metric.Points[0]
 		p[0] = float64(m.Time().Unix())
-		ms[k] = p
+		if err := p.setValue(value); err != nil {
+			return nil, fmt.Errorf("unable to extract value from Fields, %s", err.Error())
+		}
+
+		if metricType == "counter" ||
```
Comment (sparrc) [inline plugins/outputs/datadog/datadog.go]: counters should _always_ be converted to a rate? Is there ever a time where this isn't true?

**Comment by sparrc** [comment]
I think this is a good and reasonable addition to the `Metric` interface :+1:. There are certainly other plugins and outputs which may want to know the collection interval. The workaround is to cache the last time that a metric was collected (or written), but it's probably better if we have it builtin to the metric.

File: plugins/outputs/datadog/datadog.go (hunk 3 above)
Comment (wryun) [inline plugins/outputs/datadog/datadog.go]: This is consistent with the dogstatsd behaviour. Admittedly, datadog also supports a (similarly undocumented) 'counter' type, but the dogstatsd server doesn't use it. But I definitely should make a comment to this effect.

File: plugins/outputs/datadog/datadog.go (hunk 1 above)
Comment (wryun) [inline plugins/outputs/datadog/datadog.go]: Ah, thanks!

File: plugins/outputs/datadog/datadog.go (hunk 2 above)
Comment (wryun) [inline plugins/outputs/datadog/datadog.go]: go maps give back the zero of the type if it doesn't exist? Though admittedly I haven't tested this :)

**Comment by wryun** [comment]
I signed the CLA, but we're probably not going to use this for now. I may or may not find the enthusiasm to clean it up - sorry! Don't mind if you close this if you don't want it hanging around in an unfinished state.

**Pull request closed without merging**